### PR TITLE
Add Miden Assembly Formatter

### PR DESCRIPTION
**Pull request by partylikeits1983**
This PR adds the Miden Assembly formatter as a separate binary to the Miden VM repo. This tool will be integrated into the CI checks on miden-base & miden-vm. There are many instances across both repositories where masm code is incorrectly formatted. This tool is the first version of the formatter until updates are made to the assembly parser to retain inline comments in assembly code. 

Adding this basic formatter will save time in PR reviews of masm code.

This tool both formats masm code and can be run in `--check` mode so that it can be used as a CI check on PRs.

Let me know if I should remove the test masm files, this is where the majority of the diff occurs. The formatting logic is only ~250 lines of code.

### CODE REVIEW

File: formatter/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+[dependencies]
+glob = "0.3"
+clap = { version = "4.1", features = ["derive"] }
+regex = "1"
+once_cell = "1.20.3"
+tempfile = "3.17.1"
+
+[[bin]]
+name = "masm-fmt"
+path = "src/main.rs"
```
Comment (PhilippGackstatter) [inline formatter/Cargo.toml]: I think it would be good to add installation instructions (from crates.io if published and the local one) in the Readme or somewhere appropriate.

File: formatter/tests/formatter_tests.rs [hunk 2]
```diff
@@ -0,0 +1,198 @@
+use std::{fs, fs::read_to_string, path::Path};
```
Comment (PhilippGackstatter) [inline formatter/tests/formatter_tests.rs]: Not sure if there's an example/test for this, but I think the behavior in if-else statements is inconsistent. E.g. the formatter outputs this:

```
proc.test
    if.true
        push.5 add
        swap
    else
        push.6 add
        swap
    end
end
```

but also this:


```
proc.test
    if.true
        push.5 add
        swap
        # => [comment]

    else
        push.6 add
        swap
    end
end
```

The newline before else shouldn't be there imo, because it is also not added in the first case (and personally, I prefer it not being there in both cases).

**Review by PhilippGackstatter — COMMENTED** [review]
Thank you for adding this! I just want to point out a small finding.

File: formatter/tests/formatter_tests.rs (hunk 2 above)
Comment (partylikeits1983) [reply formatter/tests/formatter_tests.rs]: Can definitely update. The reasoning for this was because I hardcoded it to always have an empty line after a stack comment. 

Essentially this is the current logic:
```
# => [...
must be empty line
```

but it would be super easy to update to:
```
# => [...
if `end` statement no empty line, else add empty line
```

Comment (PhilippGackstatter) [reply formatter/tests/formatter_tests.rs]: > if `end` statement no empty line, else add empty line

I guess we'd want to avoid an empty line both when it's an `end` or an `else`, right? Maybe the logic could be "if there is a next instruction, insert empty line, otherwise don't", but not sure how easy it is?

File: Cargo.toml [hunk 3]
```diff
@@ -6,6 +6,7 @@ members = [
     "assembly",
     "assembly-syntax",
     "core",
+    "formatter",
```
Comment (bitwalker) [inline Cargo.toml]: Let's call this `miden-fmt`

File: formatter/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "formatter"
```
Comment (bitwalker) [inline formatter/Cargo.toml]: Let's name the crate `miden-fmt`, since that's the convention for new CLIs as we transition to `midenup` for toolchain management. It also aligns with the `miden-vm` executable crate.

File: formatter/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+[dependencies]
+glob = "0.3"
+clap = { version = "4.1", features = ["derive"] }
+regex = "1"
+once_cell = "1.20.3"
+tempfile = "3.17.1"
+
+[[bin]]
+name = "masm-fmt"
```
Comment (bitwalker) [inline formatter/Cargo.toml]: With the crate rename I mentioned in my other comment, the explicit name here can be removed

File: formatter/tests/expected/example1_formatted.masm
Comment (plafer) [inline formatter/tests/expected/example1_formatted.masm:1]: I'm all for adding `masm` files for tests, but these seem unnecessarily long. I basically don't know what I should be looking at for each test. A better approach would have each file show/test one (or a few related) concepts.

For example, you could have a file with effectively just a long procedure body, and test how long bodies get formatted. We should probably be able to get away with a few hundred lines of test masm for formatting.

Comment (partylikeits1983) [reply formatter/tests/expected/example1_formatted.masm:1]: Yes good point, will reduce the lines of masm test code. These were just to cast a very wide net when testing :)

File: Cargo.toml (hunk 3 above)
Comment (PhilippGackstatter) [reply Cargo.toml]: Btw. wouldn't it be slightly nicer if we didn't have `-` in executable names? I find `rustfmt` easier to type than `miden-fmt` (assuming that's the binary name). So `midenfmt` or `midenvm` as binary names, would be a bit more convenient, wdyt?

File: formatter/tests/formatter_tests.rs (hunk 2 above)
Comment (partylikeits1983) [reply formatter/tests/formatter_tests.rs]: Super easy, I updated it to match this logic.

File: formatter/tests/expected/example1_formatted.masm
Comment (partylikeits1983) [reply formatter/tests/expected/example1_formatted.masm:1]: Removed a bunch of test files. ~6100 added lines vs 1194 now.

**Comment by partylikeits1983** [comment]
Addressed nits, added some extra logic, and minimized the test masm code. Additionally opened #2072 to show how `miden-fmt` works and which changes it will make.

File: formatter/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,247 @@
+use std::{
```
Comment (PhilippGackstatter) [inline formatter/src/lib.rs:1]: Input:
```
# The slot in this component's storage layout where the public key is stored.
const.PUBLIC_KEY_SLOT=0

# The slot where the authentication configuration is stored.
const.AUTH_CONFIG_SLOT=1
```

Output:
```
# The slot in this component's storage layout where the public key is stored.

const.PUBLIC_KEY_SLOT=0

# The slot where the authentication configuration is stored.
const.AUTH_CONFIG_SLOT=1
```

That doesn't seem right - why does it insert a newline? I think it does this in every file for the first constant, but not the others.

File: miden-fmt/tests/expected/procedure_formatting_formatted.masm [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+    if.true
+        add.1
+        # => [val]
+    else
+        sub.1
+        # => [val]
+    end
+    add
+    # => [val]
```
Comment (PhilippGackstatter) [inline miden-fmt/tests/expected/procedure_formatting_formatted.masm]:
```suggestion
    if.true
        add.1
        # => [val]
    else
        sub.1
        # => [val]
    end

    add
    # => [val]
```
Perhaps it would make sense to insert newlines after `end`, _iff_ there is an instruction following?

File: miden-fmt/tests/expected/stack_operations_formatted.masm [hunk 8]
```diff
@@ -0,0 +1,34 @@
+export.test_stack_with_comments
+    # => []
+
+    push.1 push.2 push.3
+    # => [3, 2, 1]
+
+    dup.0
+    # => [3, 3, 2, 1]
+
+    swap
```
Comment (PhilippGackstatter) [inline miden-fmt/tests/expected/stack_operations_formatted.masm:10]: If I insert a comment before the `swap`, the newline will not be inserted, but I think it should be. E.g. this is the result:

```
    dup.0
    # => [3, 3, 2, 1]
    # swap the values
    swap
    # => [3, 3, 2, 1]
```

but I would want:

```
    dup.0
    # => [3, 3, 2, 1]

    # swap the values
    swap
    # => [3, 3, 2, 1]
```

We have this comment, instruction, stack state pattern a lot in miden-base, so I think formatting this as one unit would be best.

**Review by PhilippGackstatter — COMMENTED** [review]
Thanks for addressing my previous newline comment :)! Looks good!

I only looked at the output of the formatter and how it currently formats the MASM files in miden-base and left some comments based on that.

I guess it would be nice if it auto-formatted comments to be at most 100 characters wide, but probably not necessary for the first version.

File: formatter/src/lib.rs (hunk 6 above)
Comment (partylikeits1983) [reply formatter/src/lib.rs:1]: Yes, I also noticed this bug when formatting the multisig masm code, I fixed it.

File: miden-fmt/tests/expected/procedure_formatting_formatted.masm (hunk 7 above)
Comment (partylikeits1983) [reply miden-fmt/tests/expected/procedure_formatting_formatted.masm]: Great point, I resolved this in the formatter!

File: miden-fmt/tests/expected/stack_operations_formatted.masm (hunk 8 above)
Comment (partylikeits1983) [reply miden-fmt/tests/expected/stack_operations_formatted.masm:10]: Completely agree. Added a check that inserts a newline in between a stack comment and a regular comment.

**Comment by partylikeits1983** [comment]
> I guess it would be nice if it auto-formatted comments to be at most 100 characters wide, but probably not necessary for the first version.

Added auto-formatting of long comments.

File: Cargo.toml (hunk 3 above)
Comment (bitwalker) [reply Cargo.toml]: @PhilippGackstatter these executables are going to be invoked via the all-in-one `miden` CLI anyway, so the convention here is really for namespacing purposes, i.e. `miden fmt` will expect there to be a `miden-fmt` executable in the `bin` directory of the installed toolchain, but that directory may also be in the user's `PATH` if they have a need for directly working with the underlying binaries, in which case it is useful for all of our binaries to have a common naming convention.

Personally, I don't find the `-` to be all that annoying, but it entirely depends on one's keyboard layout and preferences. Since I expect most people will be using these tools via `miden` anyway, and shell aliases and/or symlinks are easy enough to create for anyone that wants to use `miden-fmt` directly and finds the full name annoying, I think sticking to the naming convention makes sense here.

File: miden-fmt/tests/unformatted/control_flow.masm
Comment (plafer) [inline miden-fmt/tests/unformatted/control_flow.masm:1]: I think this file is unintentionally formatted right? Looks identical to the formatted version. Not sure if that was already your intention, but I would make the unformatted version have basically no indentation

File: miden-fmt/tests/expected/long_comments_formatted.masm [hunk 9]
```diff
@@ -0,0 +1,12 @@
+# This is a very long comment that should be wrapped because it exceeds the 100 character limit and
+# contains multiple words that can be broken at appropriate boundaries
+export.test_procedure
+    # Another very long comment inside a procedure that should also be wrapped when it exceeds the
+    # maximum line length of 100 characters
+    push.1
+    # => [1]
```
Comment (PhilippGackstatter) [inline miden-fmt/tests/expected/long_comments_formatted.masm:7]: Thanks for adding this formatting!

One issue I found with miden-base is that this input:

```
    drop
    # => [allow_unauthorized_input_notes, allow_unauthorized_output_notes, num_auth_trigger_procs, pad(16)]
```

is reformatted to

```
    drop
    # => [allow_unauthorized_input_notes, allow_unauthorized_output_notes, num_auth_trigger_procs,

    # pad(16)]
```

Thinking about it now, I don't think it should touch stack comments at all. Formatting this one correctly, with indentation and all seems quite complicated to get right and not something the formatter should even concern itself with. So maybe it's best to only let this "wrap at 100 characters" feature apply to doc comments (`#!`) for now?

It currently also wraps [this table](https://github.com/0xMiden/miden-base/blob/f21386ee057ac235ba497052a5cf201b064948c1/crates/miden-lib/asm/kernels/transaction/lib/tx.masm#L36-L41). Ideally, there would be some way to disable formatting for that like `rusftmt::skip`. For now it would also work to only let wrapping apply to `#!` since here `#` is used.

File: miden-fmt/tests/expected/basic_formatting_formatted.masm [hunk 10]
```diff
@@ -0,0 +1,19 @@
+# Test basic formatting concepts: imports, constants, comments
```
Comment (PhilippGackstatter) [inline miden-fmt/tests/expected/basic_formatting_formatted.masm:1]: One other thing I noticed in miden-base is that this input:

```
        add
        # => [stack]
        # 2nd comment
```

is formatted as:

```
        add
        # => [stack]

        # 2nd comment
```

Which it probably shouldn't do? It seems to only insert a newline if the comment that follows the instruction is `# => [...]`. Is there some kind of detection for `# => [...]`?

**Review by PhilippGackstatter — COMMENTED** [review]
Thank you for adding comment wrapping. It works pretty well in miden-base, but it looks like there are a few edge cases (see inline comments).

File: miden-fmt/tests/expected/basic_formatting_formatted.masm (hunk 10 above)
Comment (partylikeits1983) [reply miden-fmt/tests/expected/basic_formatting_formatted.masm:1]: Ah! Yes. But the question is should these types of comments be allowed? I hardcoded the formatter to always add a new line after a stack comment. What do you think? It is possible to treat stack comments like regular comments, but stack comments are almost a part of the masm language syntax so that's why I almost treat them as their own "citizens" alongside opcodes.

Comment (PhilippGackstatter) [reply miden-fmt/tests/expected/basic_formatting_formatted.masm:1]: One example where the current strategy is not optimal is [here](https://github.com/0xMiden/miden-base/blob/6fd9a1ef93064295e973ce1b7a4df27896b4f111/crates/miden-lib/asm/kernels/transaction/lib/account_delta.masm#L332-L336). I sometimes write comments like this to make it clearer that we're destructuring a word into its constituent felts:

```
# => [KEY, VALUE0, ...]
# => [[faucet_id_prefix, faucet_id_suffix, 0, 0], [delta_amount, 0, 0, 0], ...]
```

Inserting a newline is not the end of the world, but also not ideal. Ideally, though, the formatter would leave it up to the dev to format this appropriately.

> But the question is should these types of comments be allowed?

I think they should be allowed. Thinking about it more, I would make the formatter only format what is "officially" part of the language, e.g. instructions, if, while, etc. I think devs should be able to format their comments mostly without having to consider what the formatter will do with it.

The only exception may be wrapping at 100 characters, which seems like a reasonable thing to do - many formatters do something like that - but otherwise I would treat comments just as black box strings. I do think there is a distinction between doc comments and regular ones. `rustfmt` for example, can assume that doc comments use Markdown, which is well-specified, and so the formatter could apply Markdown formatting there. The stack state comments in MASM on the other hand are not well-specified, and they are also not doc comments, and so the formatter should probably not try to format them.

Even when trying to format them, one difficulty is the different styles they appear in. In miden base, we sometimes write `# OS => [...]` and sometimes just `# => [...]`, which we both consider valid styles, but are just a miden-base local convention and should not apply to other MASM code.

Stack state comments are very pervasive, and so I totally understand why you made them "first class citizens" in the formatter, but handling them correctly may need more discussion. So, maybe for this first version of the formatter, I would only suggest to:
- Wrap doc comments at 100 characters.
- Leave non-doc comments as they were written.

That does go against what I said earlier [here](https://github.com/0xMiden/miden-vm/pull/2049#discussion_r2276458905) with regard to inserting newlines. As a consequence of the above, I guess the simplest solution would be to leave it to MASM devs to insert newlines as appropriate.

But maybe you or others have a clearer perspective on this. I think the important part is that we have a consistent and robust way of formatting things, and stack state comments are a bit tricky.

Comment (partylikeits1983) [reply miden-fmt/tests/expected/basic_formatting_formatted.masm:1]: That's true, I agree. Will make the stack comments follow the same rules as regular comments.

**Comment by bobbinth** [comment]
@partylikeits1983 - I think MASM syntax got changed quite a bit since this PR. Does it need to be refreshed? Or maybe better to close it for now?

**Comment by plafer** [comment]
Yes, this was discussed offline. @partylikeits1983 up to you if you want to take on this syntax update - if you do, please just open a new PR (or re-open this one).
